### PR TITLE
Use revocable memory in HashAggregationOperator

### PR DESCRIPTION
**Pull request by pnowojski**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/Operator.java [hunk 1]
```diff
@@ -76,11 +76,15 @@
      * After calling this method operator should revoke all reserved revocable memory.
      * As soon as memory is revoked returned future should be marked as done.
      */
-    default ListenableFuture<?> revokeMemory()
+    default ListenableFuture<?> startMemoryRevoke()
     {
         return NOT_BLOCKED;
     }
 
+    default void finishMemoryRevoke()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/Operator.java]: add javadoc

File: presto-main/src/main/java/com/facebook/presto/operator/Operator.java [hunk 2]
```diff
@@ -76,11 +76,15 @@
      * After calling this method operator should revoke all reserved revocable memory.
      * As soon as memory is revoked returned future should be marked as done.
      */
-    default ListenableFuture<?> revokeMemory()
+    default ListenableFuture<?> startMemoryRevoke()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/Operator.java]: add some doc that this method cannot touch OperatorContext from spawned threads.

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 3]
```diff
@@ -470,6 +462,33 @@ private void processNewSource(TaskSource source)
...
+                checkOperatorFinishedRevoking(operator);
+            }
+        }
+    }
+
+    private void checkOperatorFinishedRevoking(Operator operator)
+    {
+        ListenableFuture<?> future = revokingOperators.get(operator);
+        if (future.isDone()) {
+            getFutureValue(future);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: add comment that this is for propagating exception.

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java [hunk 4]
```diff
@@ -381,6 +383,23 @@ public void addInput(Page page)
     }
 
     @Override
+    public ListenableFuture<?> startMemoryRevoke()
+    {
+        if (aggregationBuilder != null) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: consider adding checkState that it is not called again before finishRevoke.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java [hunk 5]
```diff
@@ -135,11 +136,23 @@ public boolean isFull()
...
+    @Override
     public boolean isBusy()
     {
         return false;
     }
 
+    @Override
+    public CompletableFuture<?> startMemoryRevoke()
+    {
+        return null;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: `return CompletableFuture.completedFuture(null);`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 6]
```diff
@@ -132,9 +126,20 @@ private boolean hasPreviousSpillCompletedSuccessfully()
         }
     }
 
-    private boolean shouldSpill(long memorySize)
+    @Override
+    public CompletableFuture<?> startMemoryRevoke()
     {
-        return (memorySizeBeforeSpill > 0 && memorySize > memorySizeBeforeSpill);
+        if (!spillInProgress.isDone()) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: shouldn't it be

```
checkState(splillInProgress.isDone());
```

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 7]
```diff
@@ -93,27 +91,23 @@ public void processPage(Page page)
...
     }
 
     @Override
     public void updateMemory()
     {
-        aggregationMemoryContext.setBytes(hashAggregationBuilder.getSizeInMemory());
-
-        if (spillInProgress.isDone()) {
-            spillMemoryContext.setBytes(0L);
+        if (!spillInProgress.isDone()) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: shouldn't it be

```
checkState(splillInProgress.isDone());
```

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java [hunk 8]
```diff
@@ -328,12 +328,21 @@ public boolean isFinished()
     }
 
     @Override
+    public ListenableFuture<?> isBlocked()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: why is that needed?
It is handled by `Driver.isBlocked(Operator)`

I opt for removing this method from inteface.

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java [hunk 9]
```diff
@@ -176,7 +176,7 @@ public HashAggregationOperatorFactory(
             this.expectedGroups = expectedGroups;
             this.maxPartialMemory = requireNonNull(maxPartialMemory, "maxPartialMemory is null").toBytes();
             this.spillEnabled = spillEnabled;
-            this.memoryLimitBeforeSpill = requireNonNull(memoryLimitBeforeSpill, "memoryLimitBeforeSpill is null");
+            this.memoryLimitBeforeSpill = requireNonNull(memoryLimitBeforeSpill, "memoryLimitForMerge is null");
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: rename variable and argument

File: presto-main/src/test/java/com/facebook/presto/operator/OperatorAssertion.java [hunk 10]
```diff
@@ -83,6 +84,11 @@ else if (!finishing) {
                 }
             }
 
+            if (!operator.getOperatorContext().isWaitingForSystemRevocableMemory().isDone()) {
+                getFutureValue(operator.startMemoryRevoke());
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/operator/OperatorAssertion.java:70]: you may add:

``` java
interface Operator {

   default void synchronousMemoryRevoke() {
      getFutureValue(startMemoryRevoke());
      finishMemoryRevoke();
   }
} 
```

and call it here. Not sure if that is worth an effort.

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java [hunk 11]
```diff
@@ -48,8 +48,19 @@ public static TaskContext createTaskContext(Executor executor, Session session,
 
     public static TaskContext createTaskContext(Executor executor, Session session, DataSize maxMemory, DataSize preallocated)
     {
-        MemoryPool memoryPool = new MemoryPool(new MemoryPoolId("test"), new DataSize(1, GIGABYTE));
-        MemoryPool systemMemoryPool = new MemoryPool(new MemoryPoolId("testSystem"), new DataSize(1, GIGABYTE));
+        return createTaskContext(executor, session, maxMemory, preallocated, new DataSize(1, GIGABYTE), new DataSize(1, GIGABYTE));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java]: change commit message to reference TestingTaskContext

Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java]: And add commit message and/or comment to explain why is that needed.

**Review by losipiuk — CHANGES_REQUESTED** [review]

**Comment by losipiuk** [comment]
Some comments. Generally looks good. Tests are kinda hard to follow.

File: presto-main/src/main/java/com/facebook/presto/operator/Operator.java (hunk 2 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/Operator.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/Operator.java (hunk 1 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/Operator.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java (hunk 3 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java (hunk 9 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: done

Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java (hunk 4 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: It's not required in this operator. You could call `startMemoryRevoke` and `finishMemoryRevoke` arbitrary number of times in whatever order you want, so it's a bad place for such a `checkState`

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java (hunk 11 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java]: it is used in next commit

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java (hunk 8 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 7 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 6 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: done
